builder_web.ml

```diff
@@ -0,0 +1,331 @@
+    Utils.Json.
+      ( get "left_packages" json,
+        get "right_packages" json,
+        get "different_versions" json )
+  with
+  | ( Some (`List left_packages_json),
+      Some (`List right_packages_json),
+      Some (`List different_versions_json) ) ->
+      let* left_packages = list_of_json package_of_json left_packages_json in
+      let* right_packages = list_of_json package_of_json right_packages_json in
+      let* different_versions =
+        list_of_json package_version_of_json different_versions_json
+      in
+      Ok { left_packages; right_packages; different_versions }
+  | _ ->
+      Error
+        (`Msg
+          ("invalid json for builder_web env/package_diff_of_json: "
+          ^ Yojson.Basic.to_string (`Assoc json)))
+
+let opam_diff_of_json json =
+  match
+    Utils.Json.
+      ( get "opam_diff" json,
+        get "version_diff" json,
+        get "only_in_left" json,
+        get "only_in_right" json,
+        get "duniverse_diff" json )
+  with
+  | ( Some (`List o_diff_json),
+      Some (`List version_diff_json),
+      Some (`List only_in_left_json),
+      Some (`List only_in_right_json),
+      Some (`Assoc duniverse_diff_json) ) ->
+      let* opam_diff = list_of_json o_diff_of_json o_diff_json in
+      let* version_diff =
+        list_of_json package_version_of_json version_diff_json
+      in
+      let* only_in_left = list_of_json package_of_json only_in_left_json in
+      let* only_in_right = list_of_json package_of_json only_in_right_json in
+      let* duniverse_diff = duniverse_diff_of_json duniverse_diff_json in
+      Ok
+        { opam_diff; version_diff; only_in_left; only_in_right; duniverse_diff }
+  | _ ->
+      Error
+        (`Msg
+          ("invalid json for builder_web: "
+          ^ Yojson.Basic.to_string (`Assoc json)))
+
+let compare_of_json = function
+  | `Assoc xs -> (
+      match
+        Utils.Json.
+          ( get "left" xs,
+            get "right" xs,
+            get "env_diff" xs,
+            get "package_diff" xs,
+            get "opam_diff" xs )
+      with
+      | ( Some (`Assoc left_json),
+          Some (`Assoc right_json),
+          Some (`Assoc env_diff_json),
+          Some (`Assoc package_diff_json),
+          Some (`Assoc opam_diff_json) ) ->
+          let* left = build_of_json (`Assoc left_json) in
+          let* right = build_of_json (`Assoc right_json) in
+          let* env_diff = diff_of_json env_diff_json in
+          let* package_diff = diff_of_json package_diff_json in
+          let* opam_diff = opam_diff_of_json opam_diff_json in
+          Ok { left; right; env_diff; package_diff; opam_diff }
+      | _ ->
+          Error
+            (`Msg
+              ("invalid json for builder_web diff: "
+              ^ Yojson.Basic.to_string (`Assoc xs))))
+  | js ->
+      Error
+        (`Msg
+          ("invalid json for builder_web diff, expected a dict: "
+         ^ Yojson.Basic.to_string js))
+
+let send_request http_client url =
```

Comment:
should we use the `base_url` here directly? i.e. in the code below have a `let url = base_url ^ path in` where `path` is the argument here -- this would ensure that only base_url is accessed, and user code doesn't need to refer to Builder_web.base_url manually.